### PR TITLE
Remove use of Unsafe in UTF-8 formatters, plus performance improvements

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Fixes https://github.com/dotnet/corefx/issues/25648.

My first attempt to switch the implementations from unsafe code to safe buffer-based code was a bit naive and resulted in considerable performance degradation (double-digit percentage loss of throughput across many APIs). I've spent time reworking the implementations to to work around the performance loss, and in many cases the new implementations are faster than the originals.

For reviewers: I recommend looking at each commit in isolation, as each commit deals with a very specific formatter. It'll also be easier to see which helper routines in `FormatterHelpers` correlate with which implementations. This also allows individual commits to be backed out without affecting the rest of the PR if reviewers deem a particular commit as unwanted.

### CODE REVIEW

File: src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs [hunk 1]
```diff
@@ -28,21 +28,21 @@ private static bool TryFormatUInt64(ulong value, Span<byte> buffer, out int byte
                 case 'g':
                     if (format.HasPrecision)
                         throw new NotSupportedException(SR.Argument_GWithPrecisionNotSupported); // With a precision, 'G' can produce exponential format, even for integers.
-                    return TryFormatUInt64D(value, format.Precision, buffer, out bytesWritten);
+                    return TryFormatUInt64D(value, format.Precision, buffer, false /* insertNegationSign */, out bytesWritten);
```
Comment (AtsushiKan) [inline src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs]: C# now allows name-tagging arguments without name-tagging every argument after it:

 `return TryFormatUInt64D(value, format.Precision, buffer, false /* insertNegationSign */, out bytesWritten);`

==>

 `return TryFormatUInt64D(value, format.Precision, buffer, insertNegationSign: false, out bytesWritten);`

**Comment by AtsushiKan** [comment]
What are the code coverage numbers after this change?

File: src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Boolean.cs [hunk 2]
```diff
@@ -27,28 +29,77 @@ public static partial class Utf8Formatter
...
+                    if ((uint)4 >= (uint)buffer.Length)
+                    {
+                        goto BufferTooSmall;
+                    }
 
-            if (!result.TryCopyTo(buffer))
-            {
-                bytesWritten = 0;
-                return false;
+                    const uint FalsValueUppercase = ('F' << 24) + ('a' << 16) + ('l' << 8) + ('s' << 0);
```
Comment (AtsushiKan) [inline src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Boolean.cs]: Nit: "Fals" -> "False"

Comment (AtsushiKan) [reply src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Boolean.cs]: Nvm - I see the "e" doesn't fit into a uint :-)

File: src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs (hunk 1 above)
Comment (GrabYourPitchforks) [reply src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs]: Didn't know that - nifty! :)

File: src/System.Memory/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs [hunk 3]
```diff
@@ -16,111 +16,242 @@ namespace System.Buffers.Text
...
+            {
+                ThrowHelper.ThrowArgumentException_InvalidTypeWithPointersNotSupported(typeof(T));
+            }
+#else
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                ThrowHelper.ThrowArgumentException_InvalidTypeWithPointersNotSupported(typeof(T));
+            }
+#endif
+            return Span<byte>.DangerousCreate(null, ref Unsafe.As<T, byte>(ref value), Unsafe.SizeOf<T>());
```
Comment (jkotas) [inline src/System.Memory/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs]: This is not valid - it crash with slow Span. 

It is not valid to pass null as first argument to DangerousCreate,

Comment (benaadams) [reply src/System.Memory/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs]: Previous closed issue about this https://github.com/dotnet/corefx/issues/26124

File: src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs (hunk 1 above)
Comment (GrabYourPitchforks) [reply src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs]: This is now addressed.

**Comment by GrabYourPitchforks** [comment]
@AtsushiKan I don't have code coverage numbers available. I'm riding on top of the existing (quite comprehensive) unit tests for this functionality, plus I added a few extra test cases for situations where the original code produced incorrect results.

**Comment by AtsushiKan** [comment]
@GrabYourPitchforks  - You can get code coverage results by doing this:

```
   cd <repo>\src\System.Memory
   msbuild /t:build /p:DebugType=pdbonly
   cd <repo>\src\System.Memory\tests

   msbuild -t:clean;build;test -p:Coverage=true
   start <repo>\bin\tests\coverage\index.htm
```

The coverage numbers for Utf8Formatting were at a 100% when I checked in the original stuff - it's probably regressed some since then but it'd be good to know at least that the stuff you touched is at 100%.

**Comment by MattGal** [comment]
@dotnet-bot test Linux x64 Release Build please
@dotnet-bot test OSX x64 Debug Build please
@dotnet-bot test UWP CoreCLR x64 Debug Build please
@dotnet-bot test Windows x64 Debug Build please
@dotnet-bot test Windows x86 Release Build please

**Comment by GrabYourPitchforks** [comment]
@AtsushiKan It has regressed slightly (89%), but those were changes separate from this PR. Everything looks good here from a code coverage perspective. :)

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test *

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test Windows x86 Release Build
@dotnet-bot test UWP CoreCLR x64 Debug Build

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test all

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test Linux x64 Release Build
@dotnet-bot test OSX x64 Debug Build 

@dotnet-bot test UWP CoreCLR x64 Debug Build
@dotnet-bot test Windows x64 Debug Build
@dotnet-bot test Windows x86 Release Build

File: src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Boolean.cs [hunk 4]
```diff
@@ -27,28 +29,77 @@ public static partial class Utf8Formatter
...
+                    const uint TrueValueUppercase = ('T' << 24) + ('r' << 16) + ('u' << 8) + ('e' << 0);
+                    if (!BinaryPrimitives.TryWriteUInt32BigEndian(buffer, TrueValueUppercase))
+                    {
+                        goto BufferTooSmall;
+                    }
+                }
+                else if (symbol == 'l')
+                {
+                    const uint TrueValueLowercase = ('t' << 24) + ('r' << 16) + ('u' << 8) + ('e' << 0);
+                    if (!BinaryPrimitives.TryWriteUInt32BigEndian(buffer, TrueValueLowercase))
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Boolean.cs]: Would writing TryFormat using BinaryPrimitives still be necessary once TryCopyTo is optimized?
https://github.com/dotnet/coreclr/issues/15076

File: src/System.Memory/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs [hunk 5]
```diff
@@ -16,111 +16,213 @@ namespace System.Buffers.Text
...
+        /// <summary>
+        /// Fills a buffer with the ASCII character '0' (0x30).
+        /// </summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static void WriteFractionDigits(long value, int digitCount, ref byte buffer, int index)
+        public static void FillWithAsciiZeros(Span<byte> buffer)
         {
-            for (int i = FractionDigits; i > digitCount; i--)
-                value /= 10;
+            // This is a faster implementation of Span<T>.Fill().
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs]: Why is this faster? Is it due to buffer.Length being too small in this case?

Wouldn't Fill using InitBlockUnaligned be just as fast, if not faster:
https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Span.cs#L219-L231

File: src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Signed.D.cs [hunk 6]
```diff
@@ -2,74 +2,54 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
 #if !netstandard
 using Internal.Runtime.CompilerServices;
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Signed.D.cs]: Do we still need this using directive?

Comment (ahsonkhan) [reply src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Signed.D.cs]: Removed it here: https://github.com/dotnet/corefx/pull/26598

File: src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Boolean.cs (hunk 4 above)
Comment (ahsonkhan) [reply src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Boolean.cs]: cc @GrabYourPitchforks

**Pull request merged**